src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/BuildUpToDateCheck.cs

```diff
@@ -0,0 +1,227 @@
+        {
+            _projectLockService = projectLockService;
+            _projectSystemOptions = projectSystemOptions;
+            _configuredProject = configuredProject;
+            _projectLogger = projectLogger;
+            _projectItemsSchema = projectItemSchema;
+            _fileTimestampCache = fileTimestampCache;
+        }
+
+        public async Task<bool> IsUpToDateAsync(BuildAction buildAction, TextWriter logger, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (buildAction != BuildAction.Build)
+            {
+                return false;
+            }
+
+            using (var access = await _projectLockService.ReadLockAsync())
+            {
+                var project = await access.GetProjectAsync(_configuredProject, cancellationToken);
+
+                void Log(string traceMessage)
+                {
+                    _projectLogger.WriteLine($"FastUpToDate: {traceMessage}");
+                }
+
+                Log($"Starting check for project '{project.FullPath}'.");
+
+                var timestampCache = _fileTimestampCache != null ? _fileTimestampCache.Value.TimestampCache : new Dictionary<string, DateTime>(StringComparer.OrdinalIgnoreCase);
+
+                DateTime GetLatestTimestamp(List<string> paths)
+                {
+                    var latestTime = DateTime.MinValue;
+
+                    foreach (var path in paths)
+                    {
+                        if (!timestampCache.TryGetValue(path, out var time))
+                        {
+                            if (File.Exists(path))
+                            {
+                                time = File.GetLastWriteTimeUtc(path);
+                                timestampCache[path] = time;
+                                Log($"Path '{path}' had live timestamp '{time}'.");
+                            }
+                            else
+                            {
+                                Log($"Path '{path}' did not exist.");
+                            }
+                        }
+                        else
+                        {
+                            Log($"Path '{path}' had cached timestamp '{time}'.");
+                        }
+
+                        if (latestTime < time)
+                        {
+                            latestTime = time;
+                        }
+                    }
+
+                    return latestTime;
+                }
+
+                if (_projectItemsSchema == null)
+                {
+                    Log($"Skipping because '{typeof(IProjectItemSchemaService).FullName}' component could not be found.");
+                    Report.IfNotPresent(_projectItemsSchema);
+                    return false;
+                }
+
+                if (!string.IsNullOrEmpty(project.GetPropertyValue(DisableFastUpToDateCheckProperty)))
```

Comment:
Using the project properties has a some:
1. You can change the backing source of the property in the future without have to update code (maybe you want this to go into the .user file eventually?)
2. Gives you clear choices between evaluated value and unevaluated value.
3. Removes the need for your own project lock. Doesn't matter here as this is already in a lock.